Scrapers protocols

tbergmueller opened the pull request:
This PR contains the definitive architecture, code structure and implementation of OpenIbex core and chain:

### Plugin architecture
A singleton "OiPluginRegistry" is available on initialization. Each plugin is a new class, instantiated with an `@RegisterPlugin` decorator. Plugins can have dependencies and are initialized in order.
* Plugins have services. I.e. the chain instance is a service of the plugin `openibex.chain` the decorator `@WithPluginServices('openibex.chain', 'chain') makes it available as `this.chain` within any decorated class. Plugin services have an `init(plugin: OiPlugin` method that is executed when the plugin is initialized.
* Plugins have configuration: Add structure in the `config.plugins.<namespace>.<name>` config object. This config will be available as `plugin.config` or `@WithPluginServices('<namespace>.<name>', 'config').
* A generic `@OnPluginInitHook()` allows to execute an `init(plugin: OiPlugin)` method on any class.

### The chainable chain module
The chain module is an abstraction of generic blockchain functions. It can handle blocks, contracts, providers, rate limits and will be able to handle accounts, transactions and nodes as well. 

Every chain is treated through the exact same API. The chain module accepts `OiChainBundles` of `ProviderHandler, BlockHandler, ContractHandler....` which inherits from a generic handler and provides a platform specific implementation.

The chain API is chainable, i.e. `chain.contract(<assetArtifact>).getAPI(<walletname>).transfer(from, to, amount)` is a valid call.

#### Asset namespaces
According to CAIP-19 every token has an "asset namespace". We use those namespaces to identify contract types and treat contracts themselves as assets. That way the contract handler can find the abi.

Similar to chain handlers, asset handlers can be registered as well. They consist of:
* An ABI in normal JSON format.
* An API inherited from OiContractAPI
* A Connector inherited from OiContractConnector.

The `openibex.chain` plugin provides a `caip` service to ease handling of addresses.

### Protocols
Protocols consist of multiple contracts on multiple chains. The OiProtocol base class bundles those in a protocol-specific chain agnostic API and connector management. Protocols can created similar to asset namespace services and registered with the `openibex.protocols` plugin.

####

Commit:
#3 - Simplifying code: Added check for supported platforms, improved platform specific factories.

Commit:
#3 - Refactoring to enable plugins for platforms.

Commit:
#3 - Moved Ethereum specific indexer, subscriber and contracts into seperate module.

Commit:
#3 - Untangled ethers from the chain module. Chain module is now completely agnostic.

Commit:
#4 - Preliminary outline of consumer / producer architecture done during #3 cleanup.

Commit:
#4 - Added all the files and class skeletons required for data layer during #3 cleanup.

Commit:
#3 - Final cleanup, mostly docs

Commit:
Scraper works already.#4 and #5 - Added producers and consumers, started
protocol logic. Scraper works already.#4 and #5 - Added producers and
consumers, started protocol logic. Scraper works already.#4 and #5 -
Added producers and consumers, started protocol logic. Scraper works
already.

Commit:
#4 and #5: Protocol logic consumer side.

Commit:
#5 - Protocol discovery and scraper tested.

Commit:
#4 - Debugging of protools, indexer and connectors. Works fine now.

Commit:
#5 - Added protocolMaps and scraper config. CLI now supports --scrape (see help).

Commit:
#5 - Forgot to add token index.ts in last commit.

Commit:
#5 - Debugging connectors for concurrency and initializations.

Commit:
#2 - Plugin system needs improvement: Introduced Dependencies.

Commit:
#2 - Plugin Architecture implemented. Commit is broken and for sync only, @openibex/ethereum does not compile.

Commit:
#2 - Addon. Commit is broken and for sync only, @openibex/ethereum does compile but functionality is still untested.

Commit:
#2 - Plugin structure mostly done. Some issues with imports in ethereum still pending. Connect and indexing work.

Commit:
#2 - CLI changes for previous commit

Commit:
#2 - Config updated. Ethereum networks are now configured within the plugin.

Commit:
#2 - Plugin architecture finished. Including dependencies and such.

Commit:
#2 - CLI changed to new plugin architecture.

Commit:
#2 - Plugin architecture finished. Including dependencies and such.

Commit:
#4 - Re-Added support for protocol registry after #2 is finished, also improved protocol registration & handling

Commit:
#2 - Small improvement: Make plugin config available in decorator so that they're accessible as an OiPluginService.

Commit:
#2 - Wallets are integrated into the new plugin system as well. Exec works.

tbergmueller committed:
config-example: Fix intentation of plugins

tbergmueller commented on .vscode/launch.json:
I'm not entirely sure, but imho launch configurations should not go into the main repository. Code-Style etc yes, but launch-configurations of certain commands no. I'd rather add these commands as scripts, i.e. instead of the VsCode launch configuration "Initialize" add a "yarn init".

Also I'm not quite sure why we use `npx tsx packages/cli/src/index.ts init` here, when the typical way to initialize is `yarn run oi init` (where oi is build with tsup). Seems to me like a big footgun to overlook compile time errors, depending on how you work during development (via cmdline, via vscode-launch etc)

LukOfDecen replied:
Ack. That just slipped in there.

Hm.. TSX-Statement is interesting. Do we have the same package.json.

tbergmueller replied:
Deleted .vscode folder alltogether with https://github.com/openibex/openibex/commit/a7bf8eed00c7a26fd31c7c1114025a82f80f89e0

I use the package.json as on github. (I also don't use VSCODE launch configurations, I debug via cmdline and the yarn-commands we propose in README directly)

tbergmueller commented:
There is one issue that's presumably related to racing conditions. The command `yarn run oi init` frequently terminates in an error state with the following issue:

```
Error: Could not append entry:
Key "zdpuAsrbHVrXFUDJ5L1bgCbk5wzWw3rv5FW1DVqR4McaRu21U" is not allowed to write to the log
    at task (file:///home/tbergmueller/DEV/lukas/openibex/node_modules/@orbitdb/core/src/oplog/log.js:184:15)
    at async file:///home/tbergmueller/DEV/lukas/openibex/node_modules/p-queue/dist/index.js:187:36
```
Presumably, this happens when opning the openibex.values.v1 and then writing preload-vals. (I did not verify). 

It happens in around 30-50% of the cases, when I try to initialize the framework. A simple retry without any changes fixes the problem.

I'm open to moving this to a ticket for feature fixing (or tackle it with #15 ), but if it is fixable we should do it before merging, as we'd knowingly merge unstable software to main.


Complete log:
```
yarn run oi init                                                                                                                        :(
yarn run v1.22.22
$ tsx --watch packages/cli/src/index.ts init
2024-11-07 10:58:01 info: Initializing dapp, start enabled: no..
2024-11-07 10:58:01 info: Starting Helia IPFS Server...
2024-11-07 10:58:01 info: Initialized OrbitDB, identity: 02358af396ca26c6187379f06f3b8c0c5c3950548743d5f0032ba10f8c98e3aa93
2024-11-07 10:58:01 info: Peer monitoring started...
2024-11-07 10:58:01 info: Node status, PeerId: 12D3KooWKJVd5C6HHkzHtK7XqbBidRUYaTyRAQCEWj484nQtCefA connected peers: standalone
2024-11-07 10:58:02 info: Initializing Dapp with coreDB Address: /orbitdb/zdpuAsjesTBeMxSQj3pfBCt2VdKdhJKfJh4MYx2dZ1qngSuU2
@openibex/cli init

Initialize the project

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]
  --start    If run is on, the node is started straight after init. Useful for m
             emory-only config.           [choices: "yes", "no"] [default: "no"]

Error: Could not append entry:
Key "zdpuAsrbHVrXFUDJ5L1bgCbk5wzWw3rv5FW1DVqR4McaRu21U" is not allowed to write to the log
    at task (file:///home/tbergmueller/DEV/lukas/openibex/node_modules/@orbitdb/core/src/oplog/log.js:184:15)
    at async file:///home/tbergmueller/DEV/lukas/openibex/node_modules/p-queue/dist/index.js:187:36
Failed running 'packages/cli/src/index.ts init'
^C
130 tbergmueller@carl ~/DEV/lukas/openibex (git)-[scrapers-protocols] %                                                                                                                                         :(
130 tbergmueller@carl ~/DEV/lukas/openibex (git)-[scrapers-protocols] %                                                                                                                                         :(
130 tbergmueller@carl ~/DEV/lukas/openibex (git)-[scrapers-protocols] % yarn run oi init                                                                                                                        :(
yarn run v1.22.22
$ tsx --watch packages/cli/src/index.ts init
2024-11-07 10:58:10 info: Initializing dapp, start enabled: no..
2024-11-07 10:58:10 info: Starting Helia IPFS Server...
2024-11-07 10:58:11 info: Initialized OrbitDB, identity: 02b8fceb0656fb7f17a744cb052c02ee9a5ed4a682df3fd0d31f42f8a58f0d0c1d
2024-11-07 10:58:11 info: Peer monitoring started...
2024-11-07 10:58:11 info: Node status, PeerId: 12D3KooWBrNpFWY8a2kv37wmbgpKJiFzrVnsmPBmiCMKyuhd95t9 connected peers: standalone
2024-11-07 10:58:11 info: Initializing Dapp with coreDB Address: /orbitdb/zdpuB2e7mGicQhUfBqXHc3i65nwcKyJH19qYDdrrdKtrHAAKC
2024-11-07 10:58:11 info: Opening DB for openibex, values with tag "undefined" in revision 1
2024-11-07 10:58:11 info: Opened DB openibex.values.v1 at address /orbitdb/zdpuAkaDUfzFpiTiLjwJJEsv2VUSWvgcKfSd4mMyyCc5LVihj
2024-11-07 10:58:11 info: Plugin openibex.chain initializing service caip.
2024-11-07 10:58:11 info: Plugin openibex.chain initializing service resolver.
2024-11-07 10:58:11 info: Opening DB for openibex.chain, resolver with tag "undefined" in revision 1
2024-11-07 10:58:11 info: Plugin openibex.chain initializing service chain.
2024-11-07 10:58:11 info: Opened DB openibex.chain.resolver.v1 at address /orbitdb/zdpuArzgVbNKc6Qo2bT84TzmfHgxR8fKPTA6nhEJLGPkC7187
2024-11-07 10:58:11 info: Initialized plugin openibex.chain.
2024-11-07 10:58:11 info: Plugin openibex.ethereum running hook register eip155 chain.
2024-11-07 10:58:11 info: Initialized wallet default on eip155
2024-11-07 10:58:11 info: Initialized wallet alice on eip155
2024-11-07 10:58:11 info: Plugin openibex.ethereum running hook registring erc20.
2024-11-07 10:58:11 info: Registered contract erc20 on platform eip155
2024-11-07 10:58:11 info: Plugin openibex.ethereum running hook registring erc173.
2024-11-07 10:58:11 info: Registered contract erc173 on platform eip155
2024-11-07 10:58:11 info: Initialized plugin openibex.ethereum.
2024-11-07 10:58:11 info: Plugin openibex.protocols initializing service protocols.
2024-11-07 10:58:11 info: Plugin openibex.protocols running hook registring token.
2024-11-07 10:58:11 info: Registered contract usd-circle on platform eip155
2024-11-07 10:58:11 info: Registered contract usd-circle on platform solana
2024-11-07 10:58:11 info: Initialized plugin openibex.protocols.
2024-11-07 10:58:11 info: Initialized plugin openibex.usd-circle.
2024-11-07 10:58:11 info: Preload entry undefined.core.test to value: undefined
2024-11-07 10:58:11 info: Preload entry openibex.chain.lastblock to value: undefined
2024-11-07 10:58:11 info: Resolver Database is creating index, waiting for 5 seconds.





2024-11-07 10:58:16 info: stopNode: Closing database /orbitdb/zdpuB2e7mGicQhUfBqXHc3i65nwcKyJH19qYDdrrdKtrHAAKC
2024-11-07 10:58:16 info: stopNode: Closing database /orbitdb/zdpuAkaDUfzFpiTiLjwJJEsv2VUSWvgcKfSd4mMyyCc5LVihj
2024-11-07 10:58:16 info: stopNode: Closing database /orbitdb/zdpuArzgVbNKc6Qo2bT84TzmfHgxR8fKPTA6nhEJLGPkC7187
2024-11-07 10:58:16 info: stopNode: Closing OrbitDB
2024-11-07 10:58:16 info: stopNode: Stopping Orbitdb and Helia. Please wait.
```

LukOfDecen commented:
> There is one issue that's presumably related to racing conditions. The command `yarn run oi init` frequently terminates in an error state with the following issue:

Can you check this on the new cli branch?

There we have an object oriented approach to ensure the node is started. Also it's Helia 5 and libp2p 2.2.2.

However: There we still have the problem of the sometimes hanging threads.

tbergmueller committed:
Rename Oi*Register to Oi*Bundle

Following a longer discord discussion with Lukas, rename the Oi*Register to Oi*Bundle.
This is because Registers [sic! => registry] are usually filled and (re-)configured at runtime,
while these classes are rather immutable classes that form a functional unit
for e.g. a chain, a contract, a protocol etc.

Further rename related files to bundle.ts and move to top-level of each package/src

tbergmueller committed:
OiCore.getService(): Errorhandling and docs

tbergmueller committed:
core: Purge exported oiCorePlugin w/o changing functionality

For most external cases, the OiCore via async getOiCore() is available.
OiCore got a .getPlugin(namespace, pluginName), which returns fully initialized plugins.
This is more intuitive, since the average developer will not expact to receive
non-initialized plugins.

If retrieval of potentially uninitialized plugins is needed,
e.g. for other plugins, which use it in constructors and their init functions
and are aware the plugins may not be initialized yet,
OiPluginRegistry.getInstance() can be used directly.

This further adds a warning-message, when getOiCore() is called multiple times
in parallel during initialization as well as add a safe-guard to OiCore class
to ensure it can only be initialized once.

tbergmueller committed:
Remove .vscode/launch.json

tbergmueller commented:
> Can you check this on the new cli branch?

Will do, in case the issue persists, I'll address it in #18
